src/ServiceControl.Config/Commands/EditServiceControlInstanceCommand.cs

```diff
@@ -1,28 +1,36 @@
 namespace ServiceControl.Config.Commands
 {
     using System;
+    using Caliburn.Micro;
+    using ServiceControl.Config.Events;
     using ServiceControl.Config.Framework;
     using ServiceControl.Config.Framework.Commands;
     using ServiceControl.Config.UI.InstanceDetails;
     using ServiceControl.Config.UI.InstanceEdit;
     using ServiceControlInstaller.Engine.Instances;
-    
+
     class EditServiceControlInstanceCommand : AbstractCommand<InstanceDetailsViewModel>
     {
         private readonly Func<ServiceControlInstance, ServiceControlEditViewModel> editViewModel;
+        private readonly IEventAggregator eventAggregator;
         private readonly IWindowManagerEx windowManager;
 
-        public EditServiceControlInstanceCommand(IWindowManagerEx windowManager, Func<ServiceControlInstance, ServiceControlEditViewModel> editViewModel) : base(null)
+        public EditServiceControlInstanceCommand(IWindowManagerEx windowManager, Func<ServiceControlInstance, ServiceControlEditViewModel> editViewModel, IEventAggregator eventAggregator) : base(null)
         {
             this.windowManager = windowManager;
             this.editViewModel = editViewModel;
+            this.eventAggregator = eventAggregator;
         }
 
         public override void Execute(InstanceDetailsViewModel viewModel)
         {
             var editVM = editViewModel((ServiceControlInstance)viewModel.ServiceInstance);
 
             windowManager.ShowInnerDialog(editVM);
+
+            editVM.UpdateInstanceFromViewModel((ServiceControlInstance)viewModel.ServiceInstance);
+
+            eventAggregator.PublishOnUIThread(new RefreshInstances());
```

Comment:
Is there a better way to structure this? We create the edit model, show it with the instance and then update it again with the instane and the publish an event that is subscribed by another VM that just triggers property changed. Would it simplify things to send the instance over the dispatched event? Then the edit model could subscribe to that and rebind as well as the instance view model?

Reply:
Not really without reworking the UI command pattern.

When the command is invoked, a `ServiceControlEditViewModel` is created from a `ServiceControlInstance` object. This `ServiceControlInstance` object is the "truth" that needs to be updated. Once we have the `ServiceControlEditViewModel` instance, this is passed to the modal dialog. So the dialog has no reference back to the truth instance, but the command that launched the dialog does. This is why we have to call the `UpdateInstanveFromViewModel` here in order to pull the viewmodel values back into the truth object.

From here, we can bypass the eventAggregator and simply call `instance.Handle(null);` which will trigger the refresh without using the event aggregator, but that's about the only tweak I can come up with at the moment.

Any suggestions?

Reply:
Ok then I withdraw my objections. I will test this locally but from a code perspective this looks good